internal/r8/map_fetcher.go

```diff
@@ -0,0 +1,100 @@
+package r8
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const mapIndex = ".apm-android-map"
+
+var errFetcherUnvailable = errors.New("fetcher unavailable")
+
+type MapFetcher struct {
+	client *elasticsearch.Client
+	logger *logp.Logger
+}
+
+// NewMapFetcher returns a MapFetcher.
+func NewMapFetcher(c *elasticsearch.Client) *MapFetcher {
+	logger := logp.NewLogger(logs.Sourcemap)
+	return &MapFetcher{c, logger}
+}
+
+// Fetch fetches an R8 map from Elasticsearch.
+func (p *MapFetcher) Fetch(ctx context.Context, name, version string) ([]byte, error) {
+	resp, err := p.runSearchQuery(ctx, name, version)
+	if err != nil {
+		var networkErr net.Error
+		if errors.As(err, &networkErr) {
+			return nil, fmt.Errorf("failed to reach elasticsearch: %w: %v ", errFetcherUnvailable, err)
+		}
+		return nil, fmt.Errorf("failure querying ES: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// handle error response
+	if resp.StatusCode >= http.StatusMultipleChoices {
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read ES response body: %w", err)
+		}
+		if resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
+			// http.StatusNotFound -> the index is missing
+			// http.StatusForbidden -> we don't have permission to read from the index
+			// In both cases we consider the fetcher unavailable so that APM Server can
+			// fallback to other fetchers
+			return nil, fmt.Errorf("%w: %s: %s", errFetcherUnvailable, resp.Status(), string(b))
+		}
+		return nil, fmt.Errorf("ES returned unknown status code: %s", resp.Status())
+	}
+
+	body, err := io.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	decodedBody, err := base64.StdEncoding.DecodeString(string(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to base64 decode string: %w", err)
+	}
```

Comment:
I'm thinking about creating a streaming base64 decoder and zlib reader from resp.Body so that we avoid having multiple copies of large payload in memory, something like `zlib.NewReader(base64.NewDecoder(base64.StdEncoding, resp.Body))`

Reply:
Thank you, I've applied this change and it looks way cleaner now.